issue_1/test/word_registry/TestWordRegistry.ts

```diff
@@ -0,0 +1,82 @@
+import hre from "hardhat";
+import { Artifact } from "hardhat/types";
+import { SignerWithAddress } from "@nomiclabs/hardhat-ethers/dist/src/signer-with-address";
+
+import { expect } from "chai";
+import { WordRegistry } from "../../typechain/WordRegistry";
+import { Signers } from "../types";
+
+const { deployContract } = hre.waffle;
+
+const NULL_ADDRESS = '0x0000000000000000000000000000000000000000';
+
+describe("Unit tests", function () {
+  before(async function () {
+    this.signers = {} as Signers;
+
+    const signers: SignerWithAddress[] = await hre.ethers.getSigners();
+    this.signers.admin = signers[0];
+    this.signers.user1 = signers[1];
+    this.signers.user2 = signers[2];
+    console.log("admin: ", this.signers.admin.address);
+  });
+
+  describe("WordRegistry", function () {
+    beforeEach(async function () {
+      const artifact: Artifact = await hre.artifacts.readArtifact("WordRegistry");
+      this.wr = <WordRegistry>await deployContract(this.signers.admin, artifact, []);
+    });
+
+    it("should allow users to reserve words", async function () {
+      const word = "sss";
+      await expect(this.wr.connect(this.signers.user1).register(word))
+        .to.emit(this.wr, "WordRegistered").withArgs(word, this.signers.user1.address);
+    });
+
+    it("should preserve ownership of words", async function () {
+      const word = "sss";
+      expect(await this.wr.connect(this.signers.user1).wordToOwner(word))
+        .to.equal(NULL_ADDRESS);
+
+      await this.wr.connect(this.signers.user1).register(word);
+
+      expect(await this.wr.connect(this.signers.user1).wordToOwner(word))
+        .to.equal(this.signers.user1.address);
+    });
+
+    it("should allow users to unregister words", async function () {
+      const word = "sss";
+      await this.wr.connect(this.signers.user1).register(word);
```

Comment:
You can nest describe blocks:

  describe("Registry", async () => {
    beforeEach(async function () {
      const registryArtifact: Artifact = await hre.artifacts.readArtifact("Registry");
      this.registry = <Registry>await deployContract(this.signers.admin, registryArtifact);
    });

    it("User should be able to claim a name", async function () {
      await expect(this.registry.connect(this.signers.admin).claimName("this")).to.emit(this.registry, "NameClaimed");
      expect(await this.registry.nameOwnership("this")).to.equal(await this.signers.admin.getAddress());
    });

    describe("with a claimed name", async () => {
      beforeEach(async function () {
        await this.registry.connect(this.signers.admin).claimName("this")
      });

      it("Users should be able to release a name", async function () { // The name has already been claimed!
        ...
      });

      it("Users should be able to claim a second name", async function () {
       ...
      });
   });
The way you approach testing is by executing each possible action and testing all it's require, event, and state changes. Once you have done that, you open a new describe block with that action in the beforeEach and write more tests that will start from the point just after the beforeEach. The blockchain will be rolled back (kind of) after each test to the beginning of the last beforeEach.